### PR TITLE
Add padding in kv cache insertion to reduce recompilations

### PR DESCRIPTION
**Pull request by fenghuizhang**
# Description

$title.

# Tests

Manual tests and unit tests.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 1]
```diff
@@ -422,18 +422,43 @@ def insert_request_with_kv_cache(
                 "Inserting KV cache for models with multiple KV cache groups "
                 "is not supported yet.")
         block_numbers = block_ids[0]
+        num_blocks = len(block_numbers)
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py:426]: since `block_ids` is a `List[int]` now, `num_blocks` should be `len(block_ids)`?

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 2]
```diff
@@ -422,18 +422,43 @@ def insert_request_with_kv_cache(
...
+
+        # Pad the number of blocks to static bucket sizes to avoid recompilation.
+        block_buckets = [1, 2, 4, 8, 16]
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py]: `16` could be insufficient for long seq if `block_size` is small, you can just use the single value `self.max_num_blocks_per_req` to pad

File: tpu_commons/runner/jax/tpu_jax_runner.py [hunk 3]
```diff
@@ -422,18 +427,43 @@ def insert_request_with_kv_cache(
...
+        padded_num_blocks = block_buckets[bucket_index]
+        padding_size = padded_num_blocks - num_blocks
+
+        # Pad target_block_numbers. Pad with the last block_id to avoid writing
+        # to unintended blocks. The padded data in kv_cache_slices will be
+        # zeros, so this is a no-op for the valid part of the last block.
+        padded_block_numbers = jnp.pad(
+            jnp.array(block_numbers, dtype=jnp.int32),
+            (0, padding_size),
+            mode='edge',
```
Comment (xiangxu-google) [inline tpu_commons/runner/jax/tpu_jax_runner.py]: won't this be a problem? the last block will be overwritten by 0

Comment (fenghuizhang) [reply tpu_commons/runner/jax/tpu_jax_runner.py]: Fix, PTAL.

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 2 above)
Comment (fenghuizhang) [reply tpu_commons/runner/jax/tpu_jax_runner.py]: Done.

File: tpu_commons/runner/jax/tpu_jax_runner.py (hunk 1 above)
Comment (fenghuizhang) [reply tpu_commons/runner/jax/tpu_jax_runner.py:426]: The type annotation was incorrect, fixed.

**Comment by fenghuizhang** [comment]
Thanks for the comments, Xiang! Can't trust Gemini fully, especially for me who doesn't know much about these JAX functions. :p
